firedrake/variational_solver.py

```diff
@@ -35,37 +54,30 @@ def __init__(self, F, u, bcs=None, J=None,
         from firedrake import solving
         from firedrake import function
 
-        # Store input UFL forms and solution Function
-        self.F = F
-        self.Jp = Jp
-        self.u = u
         self.bcs = solving._extract_bcs(bcs)
+        self.is_linear = is_linear
+        self.Jp_eq_J = Jp is None
 
-        # Argument checking
-        if not isinstance(self.F, (ufl.Form, slate.slate.TensorBase)):
-            raise TypeError("Provided residual is a '%s', not a Form or Slate Tensor" % type(self.F).__name__)
-        if len(self.F.arguments()) != 1:
-            raise ValueError("Provided residual is not a linear form")
+        self.u = u
+        self.F = F
+        self.Jp = Jp
         if not isinstance(self.u, function.Function):
             raise TypeError("Provided solution is a '%s', not a Function" % type(self.u).__name__)
-
         # Use the user-provided Jacobian. If none is provided, derive
         # the Jacobian from the residual.
         self.J = J or ufl_expr.derivative(F, u)
 
-        if not isinstance(self.J, (ufl.Form, slate.slate.TensorBase)):
-            raise TypeError("Provided Jacobian is a '%s', not a Form or Slate Tensor" % type(self.J).__name__)
-        if len(self.J.arguments()) != 2:
-            raise ValueError("Provided Jacobian is not a bilinear form")
-        if self.Jp is not None and not isinstance(self.Jp, (ufl.Form, slate.slate.TensorBase)):
-            raise TypeError("Provided preconditioner is a '%s', not a Form or Slate Tensor" % type(self.Jp).__name__)
-        if self.Jp is not None and len(self.Jp.arguments()) != 2:
-            raise ValueError("Provided preconditioner is not a bilinear form")
+        # Argument checking
+        check_pde_args(self)
 
         # Store form compiler parameters
         self.form_compiler_parameters = form_compiler_parameters
         self._constant_jacobian = False
 
+    def __iter__(self):
+        yield self
+        yield from chain(*self.bcs)
```

Comment:
This seems like an odd choice. Normally iterating over a thing should give you back things of that type. But here we get back the problem, and then a bunch of boundary conditions.